examples/pytorch/rgcn/link_predict_mp.py

```diff
@@ -0,0 +1,1147 @@
+    tails = tails.reshape(num_chunks, neg_sample_size, hidden_dim)
+    tails = th.transpose(tails, 1, 2)
+    tmp = (heads * r).reshape(num_chunks, chunk_size, hidden_dim)
+    return th.bmm(tmp, tails)
+
+def calc_neg_head_score(heads, tails, r_emb, num_chunks, chunk_size, neg_sample_size, device=None):
+    hidden_dim = tails.shape[1]
+    r = r_emb
+    if device is not None:
+        r = r.to(device)
+        heads = heads.to(device)
+        tails = tails.to(device)
+    heads = heads.reshape(num_chunks, neg_sample_size, hidden_dim)
+    heads = th.transpose(heads, 1, 2)
+    tmp = (tails * r).reshape(num_chunks, chunk_size, hidden_dim)
+    return th.bmm(tmp, heads)
+
+def get_loss(h_emb, t_emb, nh_emb, nt_emb, r_emb,
+    num_chunks, chunk_size, neg_sample_size):
+    # triplets is a list of data samples (positive and negative)
+    # each row in the triplets is a 3-tuple of (source, relation, destination)
+    pos_score = calc_pos_score(h_emb, t_emb, r_emb).view(-1, 1)
+    t_neg_score = calc_neg_tail_score(h_emb, nt_emb, r_emb, num_chunks, chunk_size, neg_sample_size)
+    h_neg_score = calc_neg_head_score(nh_emb, t_emb, r_emb, num_chunks, chunk_size, neg_sample_size)
+    #pos_score = F.logsigmoid(pos_score)
+    h_neg_score = h_neg_score.view(-1, 1)
+    t_neg_score = t_neg_score.view(-1, 1)
+    #h_neg_score = F.logsigmoid(-h_neg_score).mean(dim=1)
+    #t_neg_score = F.logsigmoid(-t_neg_score).mean(dim=1)
+
+    score = th.cat([pos_score, h_neg_score, t_neg_score])
+    label = th.cat([th.full((pos_score.shape), 1.0),
+                    th.full((h_neg_score.shape), 0.0),
+                    th.full((t_neg_score.shape), 0.0)]).to(score.device)
+    predict_loss = F.binary_cross_entropy_with_logits(score, label)
+    #pos_score = pos_score.mean()
+    #h_neg_score = h_neg_score.mean()
+    #t_neg_score = t_neg_score.mean()
+    #predict_loss = -(2 * pos_score + h_neg_score + t_neg_score) / 2
+
+    return predict_loss
+
+class NodeSampler:
+    """ Used by Inference
+    """
+    def __init__(self, g, fanouts=[None]):
+        self.g = g
+        self.fanouts = fanouts
+
+    def sample_nodes(self, seeds):
+        seeds = th.tensor(seeds).long()
+        seeds = seeds.squeeze()
+
+        curr = seeds
+        blocks = []
+        for fanout in self.fanouts:
+            if fanout is None or fanout == -1:
+                frontier = dgl.in_subgraph(self.g, curr)
+            else:
+                frontier = dgl.sampling.sample_neighbors(self.g, curr, fanout)
+            block = dgl.to_block(frontier, curr)
+            curr = block.srcdata[dgl.NID]
+            blocks.insert(0, block.int())
+
+        return seeds, blocks
+
+class LinkPathSampler:
+    """ Path based sampler.
+
+    In Path sampler, we randomly select K positive edges (e.g., 30000) to construct
+    a subgraph and add corresponding reverse edges. Then we randomly remove half of
+    positive edges in order to over information leakage.
```

Comment:
Is this a standard approach? Look weird to me esp. the removing half of the positive edges part.

"in order to over ..." <--------- miss a word?